### ISSUE TITLE
How to easily reproduce the session when an error occured

### ISSUE THREAD

**Issue by kaidatavis** [Vis4Sense/SenseMapExtension#140]
This would be important to report and reproduce bug. 

It probably involves saving the browsing session, i.e., all the pages that were opened and the order/timing they were opened?

Maybe a 'playback' feature to reply the process?

**Comment by kaidatavis** [comment]
> It probably involves saving the browsing session, i.e., all the pages that were opened and the order/timing they were opened?

Will it be enough to just save the pages array (`hmPages`)?

> Maybe a 'playback' feature to reply the process?

Automatically add one page from the save session to the historyMap following the temporal order?

**Comment by kaidatavis** [comment]
Need a 'save' button on the historyMap page?

**Comment by Yuu-Han** [comment]
> This would be important to report and reproduce bug.

I agree. It is also useful for users to review the sensemaking session.

> Will it be enough to just save the pages array (`hmPages`)?

It should be enough if the input of the tree is `hmPages`. But I'm not sure how the 'revisiting previous node' action is recorded. In the current mv3, clicking a node on the tree generates a new node, but clicking existing tabs in the browser does not update the array? This might depend on the decision of #135. If choosing option 2, maybe need another status array (or adding attributes to `hmPage` object)?

**Comment by kaidatavis** [comment]
@Yuu-Han, do you want to give this a go? Maybe just a simplest version that adds a 'save' and 'load' button to the historymap (`hm.html`): 
- The 'save' button saves the current `hmPages` as a json file.
- The 'load' button loads a json file, i.e. a `hmPage` array, and then adds one `hmPage` a time to an empty 'hmPages`. We don't want to mix with the pages currently opened in the browser, as this is mainly for debugging. Also, there would be nice to have some delay before adding a new `hmPage` so user can catch up on what is happening. 
This can get more complicated by having 'play', 'pause', or even 'rewind' functions, but we can do this later.

Let me know if there is any question.

**Comment by Yuu-Han** [comment]
Yes, I will have a try.

Are we going to open a new window to display the saved session? Another question is when a new `hmPage` is added, will the page be opened in the browser or just shown as the node in the historymap?

**Comment by Yuu-Han** [comment]
It is not quite relevant to reproducing for debugging and not a high priority but I frequently find information helpful for another task when searching information for the task at hand. I don't quite know what is the term for this. Maybe something close to cross-task sensemaking? It will be nice if these links between sessions can be captured (in the distant future)?

**Comment by kaidatavis** [comment]
> Are we going to open a new window to display the saved session? 

Either is fine. We can just use your implementation. 

> Another question is when a new `hmPage` is added, will the page be opened in the browser or just shown as the node in the historymap?

Just the nodes. In the old extension user can click on a node to open that webpage in the browsers (not just for saved sessions). Opening all the pages in a saved session might create too many tabs.

**Comment by kaidatavis** [comment]
> It is not quite relevant to reproducing for debugging and not a high priority but I frequently find information helpful for another task when searching information for the task at hand. I don't quite know what is the term for this. Maybe something close to cross-task sensemaking? It will be nice if these links between sessions can be captured (in the distant future)?

That's a very good idea. Do you want to create a new issue for this, so we don't forget it? 

Another feature in the old extension is that use can highlight text or image a webpage to mark/save them. Once this is re-implemented, maybe we can introduce something like 'tags' to group content/information (similar to hashtag in x/twitter)?

**Comment by Yuu-Han** [comment]
- [ ] Putting the save and load functionality into the popup https://github.com/Vis4Sense/HistoryMap/issues/145#issuecomment-1882105990

**Comment by Yuu-Han** [comment]
For the save and load button, are we better to put it in the popup or on the historymap?

**Comment by kaidatavis** [comment]
Personally I think it is easier for the user to have the save/load on the HistoryMap page, but we can discuss at the next meeting.

**Comment by kaidatavis** [comment]
Currently the `hmPages` does not have all the information needed to reproduce the entire process, which may need to be added later (a separate issue?).

Also, it will be useful to record the time that a tab is 'active' or 'visible' (another issue?).
